### PR TITLE
feat: modal on sso navigation

### LINKED ISSUE #640: Navigating during SSO Registration should not be possible
When completing an account registration during the SSO process, the user can click on any header buttons (sign in, categories etc) and implicitly abort the SSO process, causing some illegal states and disrupting the sign in process.

**Expected Behavior**

Either navigating with an incompleted registration form should cancel the registration (like the cancel button does) or the navigation should not be possible (like it is during checkout).

**Steps to Reproduce the Bug**

- Start SSO with Auth0
- Create new customer
- On the "Complete your Registration" page, click somewhere else
- Try signing in again

### PR DESCRIPTION
**Pull request by MaxKless**
## PR Type

[x] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no API changes)
[ ] Build-related changes
[ ] CI-related changes
[ ] Documentation content changes
[ ] Application / infrastructure changes
[ ] Other: 

## What Is the Current Behavior?
Currently, you can navigate away from the registration page during sso account completion and break the process.

Issue Number: Closes #640

## What Is the New Behavior?
Now, there exists a modal that asks you to confirm whether you want to leave the sso account completion. Doing this will log you out and ensure a proper state.

## Todo
- [x] remove store stuff from registration configuration to account facade

## Does this PR Introduce a Breaking Change?

[ ] Yes
[x] No

### CODE REVIEW

**Comment by MaxKless** [comment]
Hi @marschmidt89 , please provide translations / proper text for the new translation keys:
 "account.register.confirm_leave_modal.body"
  "account.register.confirm_leave_modal.cancel_button"
  "account.register.confirm_leave_modal.confirm_button"
  "account.register.confirm_leave_modal.title"

**Review by M-Behr — CHANGES_REQUESTED** [review]
Hi Max, I have some feedback regarding the modal:

- [ ]  I would like to change the text a little bit, e.g. remove punctuation in heading and shorter button names to avoid breaks. See attachment.
- [x] The secondary button has incorrect styling. (I think it should be: class="btn btn-secondary")
- [ ] The button order does not follow our pattern.
- [ ] The form in the background moves, if I triggered cancellation via click on logo. See here:
background

Here is my suggestion for this modal:
IAD_CancelSSOREgistration

**Review by M-Behr — APPROVED** [review]
If you could hide the error message when changing the language, that would be great. Otherwise I'll write an extra ticket for it. The modal looks good. Translations should be adjusted again. Thank you!

File: src/app/pages/registration/registration-page.component.ts [hunk 1]
```diff
@@ -36,30 +33,22 @@ export class RegistrationPageComponent implements OnInit {
...
-            this.model
-          ))
-      ),
-      map(config => this.registrationFormConfiguration.getRegistrationFormConfiguration(config))
+    this.error$ = this.registrationFormConfiguration.getErrorSources();
+
+    const snapshot = this.route.snapshot;
+    this.model = this.extractModel(snapshot);
+    this.registrationConfig = this.extractConfig(snapshot);
+    this.options = this.registrationFormConfiguration.getRegistrationFormConfigurationOptions(
```
Comment (jometzner) [inline src/app/pages/registration/registration-page.component.ts:47]: Kind of lengthy, wordy and redundant. Could you rename the method into just `getOptions()`?

File: src/app/pages/registration/registration-page.component.ts [hunk 2]
```diff
@@ -79,9 +68,17 @@ export class RegistrationPageComponent implements OnInit {
...
+  extractModel(route: ActivatedRouteSnapshot) {
+    return Object.keys(route.queryParams)
+      .filter(key => !['userid'].includes(key))
+      .reduce((acc, key) => ({ ...acc, [key]: route.queryParams[key] }), { ...this.model });
+  }
+
+  extractConfig(route: ActivatedRouteSnapshot) {
+    return {
+      sso: !!route.url.find(segment => segment.path.includes('sso')),
+      userId: route.queryParams.userid,
```
Comment (jometzner) [inline src/app/pages/registration/registration-page.component.ts:80]: userid query param is now mandatory for all routes that include `sso`, am I right? Should this be documented?

File: src/app/pages/registration/registration-page.guard.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+@Injectable()
+export class RegistrationPageGuard implements CanDeactivate<RegistrationPageComponent> {
+  constructor(private registrationConfigurationService: RegistrationFormConfigurationService) {}
+
+  canDeactivate(
+    component: RegistrationPageComponent,
+    currentRoute: ActivatedRouteSnapshot
+  ): boolean | UrlTree | Observable<boolean | UrlTree> | Promise<boolean | UrlTree> {
+    return this.registrationConfigurationService.canDeactivate(component.extractConfig(currentRoute));
```
Comment (jometzner) [inline src/app/pages/registration/registration-page.guard.ts:17]: I like the simplicity of this. Nice work.

**Review by jometzner — CHANGES_REQUESTED** [review]

File: src/app/core/facades/account.facade.ts [hunk 4]
```diff
@@ -207,4 +212,22 @@ export class AccountFacade {
...
+
+  ssoRegistrationError$ = this.store.pipe(select(getSsoRegistrationError));
+  ssoRegistrationCancelled$ = this.store.pipe(select(getSsoRegistrationCancelled));
+  ssoRegistrationRegistered$ = this.store.pipe(select(getSsoRegistrationRegistered));
+
+  cancelRegistration() {
+    this.store.dispatch(cancelRegistration());
+  }
+
+  setRegistrationInfo(registrationInfo: {
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply src/app/core/facades/account.facade.ts:226]: The type existed, I just forgot about it: SsoRegistrationType

File: src/app/core/services/registration-form-configuration/registration-form-configuration.service.ts [hunk 5]
```diff
@@ -150,7 +150,38 @@ export class RegistrationFormConfigurationService {
...
+        this.accountFacade.ssoRegistrationRegistered$,
+      ]).pipe(
+        take(1),
+        switchMap(([cancelled, registered]) =>
+          iif(
+            () => cancelled || registered,
+            of(true),
+            of({}).pipe(
+              map(() => this.ngbModal.open(ConfirmLeaveModalComponent)),
+              switchMap(modalRef => race(modalRef.dismissed.pipe(mapTo(false)), from(modalRef.result))),
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply src/app/core/services/registration-form-configuration/registration-form-configuration.service.ts:171]: the modalRef delivers whatever we define for the close method in the modal itself. My implementation works only with booleans but you could put anything in there.

File: src/app/core/identity-provider/auth0.identity-provider.ts [hunk 6]
```diff
@@ -124,7 +123,7 @@ export class Auth0IdentityProvider implements IdentityProvider {
                     )
                   ),
                   switchMap((navigated: boolean) =>
-                    navigated
+                    navigated || navigated === null
```
[the remark this answers is not in the document]
Comment (MaxKless) [reply src/app/core/identity-provider/auth0.identity-provider.ts:126]: It's null if we refresh the registration page during the sso process. Usually I would test for undefined or null as `!navigated` but here I'm just trying to catch one specific case.

File: src/app/core/services/registration-form-configuration/registration-form-configuration.service.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jometzner) [reply src/app/core/services/registration-form-configuration/registration-form-configuration.service.ts]: Right, I've found that answer yesterday once I came across your changes to `confirm-leave-modal`

File: src/app/pages/registration/registration-page.component.ts [hunk 7]
```diff
@@ -26,40 +23,27 @@ export class RegistrationPageComponent implements OnInit {
...
-      ),
-      map(config => this.registrationFormConfiguration.getRegistrationFormConfiguration(config))
-    );
+    this.error$ = this.registrationFormConfiguration.getErrorSources();
+
+    const snapshot = this.route.snapshot;
+    this.model = this.registrationFormConfiguration.extractModel(snapshot);
+    this.registrationConfig = this.registrationFormConfiguration.extractConfig(snapshot);
+    this.options = this.registrationFormConfiguration.getOptions(this.registrationConfig, this.model);
+    this.fields = this.registrationFormConfiguration.getRegistrationFormConfiguration(this.registrationConfig);
```
Comment (jometzner) [inline src/app/pages/registration/registration-page.component.ts:46]: Somehow this is still bothering me. It's only naming! But I would prefer to read this line like this:
```
this.fields = this.registrationFormConfiguration.getFields(this.registrationConfig);
```

**Review by jometzner — CHANGES_REQUESTED** [review]

**Review by jometzner — APPROVED** [review]

**Pull request merged**